core/src/main/java/io/questdb/network/IODispatcherLinux.java

```diff
@@ -45,66 +53,221 @@ public void close() {
+                final long eventId = nextEventId();
+                LOG.debug().$("registering suspend event [fd=").$(fd)
+                        .$(", op=").$(operation)
+                        .$(", eventId=").$(eventId)
+                        .$(", suspendedOpId=").$(opId)
+                        .$(", deadline=").$(suspendEvent.getDeadline()).I$();
+
+                int eventRow = pendingEvents.addRow();
+                pendingEvents.set(eventRow, EVM_ID, eventId);
+                pendingEvents.set(eventRow, EVM_OPERATION_ID, opId);
+                pendingEvents.set(eventRow, EVM_DEADLINE, suspendEvent.getDeadline());
+
+                if (epoll.control(suspendEvent.getFd(), eventId, EpollAccessor.EPOLL_CTL_ADD, EpollAccessor.EPOLLIN) < 0) {
+                    LOG.error().$("epoll_ctl add suspend event failure [id=").$(eventId)
+                            .$(", err=").$(nf.errno()).I$();
+                }
+            }
         }
+        return useful;
+    }
 
-        if (offset > 0) {
-            LOG.debug().$("reg").$();
+    private void processSuspendEventDeadlines(long timestamp) {
+        int count = 0;
+        for (int i = 0, n = pendingEvents.size(); i < n && pendingEvents.get(i, EVM_DEADLINE) < timestamp; i++, count++) {
+            final long eventId = pendingEvents.get(i, EVM_ID);
+            final long opId = pendingEvents.get(i, EVM_OPERATION_ID);
+            final int pendingRow = pending.binarySearch(opId, OPM_ID);
+            if (pendingRow < 0) {
+                LOG.error().$("internal error: failed to find operation for expired suspend event [id=").$(opId).I$();
+                continue;
+            }
+            // First, remove the suspend event from the epoll interest list.
+            final C context = pending.get(pendingRow);
+            final int operation = (int) pending.get(pendingRow, OPM_OPERATION);
+            final SuspendEvent suspendEvent = context.getSuspendEvent();
+            assert suspendEvent != null;
+            if (epoll.control(suspendEvent.getFd(), eventId, EpollAccessor.EPOLL_CTL_DEL, 0) < 0) {
+                LOG.error().$("epoll_ctl remove suspend event failure [eventId=").$(eventId)
+                        .$(", err=").$(nf.errno()).I$();
+            }
+            // Next, resume the original operation and close the event.
+            resumeOperation(context, opId, operation);
+        }
+        pendingEvents.zapTop(count);
+    }
+
+    private void resumeOperation(C context, long id, int operation) {
+        // to resume a socket operation, we simply re-arm epoll
+        if (
+                epoll.control(
+                        context.getFd(),
+                        id,
+                        EpollAccessor.EPOLL_CTL_MOD,
+                        operation == IOOperation.READ ? EpollAccessor.EPOLLIN : EpollAccessor.EPOLLOUT
+                ) < 0
+        ) {
+            LOG.error().$("epoll_ctl operation mod failure [id=").$(id)
```

Comment:
should be critical

Reply:
Changed in bbec37baf832db9a846d04e59ab124f8bb84eadc